Add option for returning from LHEweight parsing if not valid xml

dteague opened the pull request:
# Overview
Setter function with variable to decide if to fail if XML is not valid

## Limitations
This only is checking if the xml fails the `isConsistent` function which checks if the XML has swapped tags (ie the tags are each at the right level). This can be expanded if needed (the other check done on the XML is if &gt; is used in lieu of >, but that seems fairly trivial). 

Also, the error handling is not very fleshed out. It prints off a stock error message to `cerr` (is there a CMSSW error logging functionality or?) and just returns, so the code does not fail. If we want to have the code fail more upstream, we could change the parsing function to return a bool with the status of the parsing.

kdlong commented:
I think this needs to involve the xml parser and determine whether or not the header text block is valid xml

dteague commented:
https://github.com/kdlong/cmssw/blob/b20316877d15cdc955bae7b0b77654945c50a956/GeneratorInterface/Core/src/LHEWeightHelper.cc#L31-L38
This is this snippet already in the code that is already run whether the flag is set or not. The thing I am worried about though is that tinyxml2 is fairly error prone: When I've run the code on files with mixed up xml tags, it usually doesn't complain even if the tags aren't balanced. So I'd think the xml parsing is only really going to pick out really fatal errors...

kdlong commented:
Right, you can leave this function you've added to check isConsistent, but I'd also change this return to an exception in the case where failIfInvalidXML_ = true.

Would also be good to add parameter set args to the LHEWeightsProduct producer to set this flag.

dteague commented:
Added a flag to the python cff. Idk how to make an optional flag using the `ParameterSet` object, so the `failIfInvalidXML` is required, which we might not want to have. Let me know if this implementation is more in line with your thoughts of how it should be!

kdlong commented:
Looks good! You can make it optional with cms.untracked.bool and iConfig.getUntrackedParameter

dteague commented:
Rebased and fixed option to be an optional parameter. I default it to false (ie don't fail if bad xml), but I can change that real quick if we want to make it default true!

kdlong commented:
Thanks, I didn't test this myself but I'm going to merge assuming it won't break anything
